diff --git a/spacewalk-koan.spec b/spacewalk-koan.spec
index 50466ac..b0286c6 100644
--- a/spacewalk-koan.spec
+++ b/spacewalk-koan.spec
@@ -12,13 +12,22 @@
 Summary: Support package for spacewalk koan interaction
 Name: spacewalk-koan
 Version: 2.8.6
-Release: 5%{?dist}
+Release: 6%{?dist}
 License: GPLv2
 Source0: https://github.com/spacewalkproject/spacewalk/archive/%{name}-%{version}.tar.gz
 URL:            https://github.com/spacewalkproject/spacewalk
 BuildArch:      noarch
 Requires:       %{pythonX}-%{name} = %{version}-%{release}
 Requires:       koan
+# dd, du
+Requires:       coreutils
+Requires:       cpio
+Requires:       e2fsprogs
+# gzip, zcat
+Requires:       gzip
+Requires:       tar
+# mount, umount
+Requires:       util-linux
 Requires:       xz
 Conflicts: rhn-kickstart
 Conflicts: rhn-kickstart-common
@@ -104,6 +113,9 @@ make -f Makefile.spacewalk-koan install PREFIX=$RPM_BUILD_ROOT ROOT=%{python3_si
 %endif
 
 %changelog
+* Tue May 21 2019 Michael Mraka <michael.mraka@redhat.com> 2.8.6-6
+- Resolves: #1703706 - require commands we use in merge-rd.sh
+
 * Tue Mar 20 2018 Tomas Kasparek <tkasparek@redhat.com> 2.8.6-5
 - don't build python2 subpackages on systems with default python3
   (tkasparek@redhat.com)
